[WIP] RD Training (#57521)
odoo/odoo: doc/howtos/rdtraining/08_relations.rst

    - Create the ``estate.property.type`` model and add the following field:

    ========================= ========================= =========================
    Field                     Type                      Attributes
    ========================= ========================= =========================
    name                      Char                      required
    ========================= ========================= =========================

    - Add the menus as displayed in the **Goal**
    - Add the field ``property_type_id`` on your ``estate.property`` model and its form, tree
      and search views

    This excercise is a good recap of the previous chapters: you need to create a
    :ref:`model <howto/rdtraining/04_basicmodel>`, set the
    :ref:`model <howto/rdtraining/05_securityintro>`, add an
    :ref:`action and a menu <howto/rdtraining/06_firstui>`, then
    :ref:`create a view <howto/rdtraining/07_basicviews>`.

    Tip: do not forget to import any new Python file in ``__init__.py``, add new data files in
    ``__manifest.py__``  and add the access rights ;-)

Once again, restart the server and refresh to see the results!

> In the real estate module, there are still two missing information we want on a property: the buyer

Reviewer: ```suggestion
In the real estate module, there are still two missing pieces of information we want on a property: the buyer
```

and the salesperson. The buyer can be any individual, but on the other end the salesperson must be
an employee of the real estate agency (i.e. an Odoo user).

In Odoo, there are two models to which we commonly refer to:

- ``res.partner``: a partner is a physical or legal entity. It can be a company, an individual, or
  even a contact address.
- ``res.users``: it contains the users of the system. Users can be 'internal': they have
  access to the Odoo backend. Or they can be 'portal': they cannot access the backend, only the
  frontend (e.g. to access their previous orders on the eCommerce).

.. exercise:: Add the buyer and the salesperson

    Add a buyer and a salesperson on the ``estate.property`` model using the two common models

>     mentioned. They should be added in a new tab of the form view, as depicted.

Reviewer: ```suggestion
    mentioned. They should be added in a new tab of the form view, as depicted in the **Goal** section.
```

**Reference**: the documentation related to this topic can be found in
:class:`~odoo.fields.Many2many`.

.. note::

    **Goal**: at the end of this section:
    
    - a new ``estate.property.tag`` model is created with the corresponding menu and action.

    .. image:: 08_relations/media/property_tag.png
        :align: center
        :alt: Property tag

    - Tags are added to the ``estate.property`` model:

    .. image:: 08_relations/media/property_many2many.png
        :align: center
        :alt: Property

In our real estate module, we want to define the concept of property tag. A property tag
is, for example, a property which is 'cozy' or 'renovated'.

A property can have **many** tags, and a tag can be assigned to **many** properties:
we will use the **many2many** concept.

A many2many is a bidirectional multiple relationship, any record on one side can be related to any
number of records on the other side.  For example, in order to define a link to the
``account.tax`` model on our test model, we could write::

    tax_ids = fields.Many2many("account.tax", string="Taxes")

By convention, many2many fields have the ``_ids`` suffix. This means that several taxes can be

> added to our test model. It behaves as a list of records, meaning that accessing the data must be

Reviewer: maybe here it could be mentioned that a list of records is known as a `recordset`

done in a loop::

    for tax in my_test_object.tax_ids:
        print(tax.name)

.. exercise:: Add the Real Estate Property Tag table

    - Create the ``estate.property.tag`` model and add the following field:

    ========================= ========================= =========================
    Field                     Type                      Attributes
    ========================= ========================= =========================
    name                      Char                      required
    ========================= ========================= =========================

    - Add the menus as displayed in the **Goal**

>     - Add the field ``tag_ids`` on your ``estate.property`` model and it its form and tree views

Reviewer: ```suggestion
    - Add the field ``tag_ids`` on your ``estate.property`` model and in its form and tree views
```

    Tip: in the view, use the ``widget="many2many_tags"`` attribute as done
    `here <https://github.com/odoo/odoo/blob/5bb8b927524d062be32f92eb326ef64091301de1/addons/crm_iap_lead_website/views/crm_reveal_views.xml#L36>`__.

>     We well cover :ref:`later <howto/rdtraining/12_sprinkles>` the ``widget`` attribute in more
>     details. For now, you can try to add or remove it and see the result ;-)

Reviewer: ```suggestion
    We will cover :ref:`later <howto/rdtraining/12_sprinkles>` the ``widget`` attribute in more
    detail. For now, you can try to add or remove it and see the result ;-)
```

One2many
========

**Reference**: the documentation related to this topic can be found in
:class:`~odoo.fields.One2many`.

.. note::

    **Goal**: at the end of this section:
    
    - A new ``estate.property.offer`` model is created with the corresponding form and tree view.
    - Offers are added to the ``estate.property`` model:

    .. image:: 08_relations/media/property_offer.png
        :align: center
        :alt: Property offers

In our real estate module, we want to define the concept of property offer. A property offer
is an amount a potential buyer offers to the seller. The offer can be lower or higher than the
expected price.

An offer applies to **one** property, but the same property can have **many** offers:
the concept of **many2one** appears once again. However, in this case we want to display the list
of offers for a given property: we will use the **one2many** concept.

A one2many virtual relationship, inverse of a many2one. For example, we defined on our test model

> a link to the ``res.partner`` model thanks to the field ``partner_id``. We can defined the inverse

Reviewer: ```suggestion
a link to the ``res.partner`` model thanks to the field ``partner_id``. We can define the inverse
```

relation, i.e. the list of test models linked to our partner::

    test_ids = fields.One2many("test.model", "partner_id", string="Tests")

The first parameter is called the ``comodel``, while the second parameter is the field we want to
inverse.

By convention, one2many fields have the ``_ids`` suffix. It behaves as a list of records, meaning
that accessing the data must be done in a loop::

    for test in partner.test_ids:
        print(test.name)

.. danger::

    Because a :class:`~odoo.fields.One2many` is a virtual relationship,

>     there *must* be a :class:`~odoo.fields.Many2one` field

Reviewer: ```suggestion
    there *must* be a :class:`~odoo.fields.Many2one` field defined in the comodel.
```

.. exercise:: Add the Real Estate Property Offer table

    - Create the ``estate.property.offer`` model and add the following fields:

    ========================= ================================ ============= =================
    Field                     Type                             Attributes    Values
    ========================= ================================ ============= =================
    price                     Float
    status                    Selection                        no copy       Accepted, Refused
    partner_id                Many2one (``res.partner``)       required
    property_id               Many2one (``estate.property``)   required
    ========================= ================================ ============= =================

    - Create a tree and form view with the ``price``, ``partner_id`` and ``status`` fields. No
      need for an action nor a menu.
    - Add the field ``offer_ids`` on your ``estate.property`` model and in its form view as

>       depicted.

Reviewer: ```suggestion
      depicted in the **Goal**.
```

There are several important things to notice here. First, we don't need an action or a menu for all
models. Some models are intended to be accessed only through another one. This is the case in our
exercise: an offer is always accessed through a property.

Second, despite the fact that the ``property_id`` field is required, we did not include it in the

> views. How comes the Odoo knows to which property our offer is linked to? Well that's part of the

Reviewer: ```suggestion
views. How come Odoo knows to which property our offer is linked to? Well that's part of the
```

  frontend (e.g. to access their previous orders on the eCommerce).

.. exercise:: Add the buyer and the salesperson

    Add a buyer and a salesperson on the ``estate.property`` model using the two common models
    mentioned. They should be added in a new tab of the form view, as depicted in the **Goal**
    section.

    The default value for the salesperson must be the current user. The buyer should not be copied.

    Tip: to get the default value, check the note below or find an example
    `here <https://github.com/odoo/odoo/blob/5bb8b927524d062be32f92eb326ef64091301de1/addons/crm/models/crm_lead.py#L92>`__.

.. note::

    The object ``self.env`` gives access to request parameters and other useful
    things:

    - ``self.env.cr`` or ``self._cr`` is the database *cursor* object; it is
      used for querying the database
    - ``self.env.uid`` or ``self._uid`` is the current user's database id
    - ``self.env.user`` is the current user's record
    - ``self.env.context`` or ``self._context`` is the context dictionary
    - ``self.env.ref(xml_id)`` returns the record corresponding to an XML id

>     - ``self.env[model_name]`` returns an instance of the given model

Author: The exercise just above:
> The default value for the salesperson must be the current user.

It also links to a part of code with `self.env`

number of records on the other side.  For example, in order to define a link to the
``account.tax`` model on our test model, we could write::

    tax_ids = fields.Many2many("account.tax", string="Taxes")

By convention, many2many fields have the ``_ids`` suffix. This means that several taxes can be
added to our test model. It behaves as a list of records, meaning that accessing the data must be
done in a loop::

    for tax in my_test_object.tax_ids:
        print(tax.name)

A list of records is known as a *recordset*, i.e. an ordered collection of records. It supports
the standard Python operations on collections, like ``len()`` and ``iter()``, plus extra set
operations like ``recs1 + recs2``.

.. exercise:: Add the Real Estate Property Tag table

    - Create the ``estate.property.tag`` model and add the following field:

    ========================= ========================= =========================
    Field                     Type                      Attributes
    ========================= ========================= =========================
    name                      Char                      required

>     ========================= ========================= =========================

Author: doesn't seem working. Not a big deal, though

on our test model a link to the ``res.partner`` model thanks to the field ``partner_id``.
We can define the inverse relation, i.e. the list of test models linked to our partner::

    test_ids = fields.One2many("test.model", "partner_id", string="Tests")

The first parameter is called the ``comodel``, while the second parameter is the field we want to
inverse.

By convention, one2many fields have the ``_ids`` suffix. They behave as a list of records, meaning
that accessing the data must be done in a loop::

    for test in partner.test_ids:
        print(test.name)

.. danger::

    Because a :class:`~odoo.fields.One2many` is a virtual relationship,
    there *must* be a :class:`~odoo.fields.Many2one` field defined in the comodel.

.. exercise:: Add the Real Estate Property Offer table

    - Create the ``estate.property.offer`` model and add the following fields:

    ========================= ================================ ============= =================
    Field                     Type                             Attributes    Values
    ========================= ================================ ============= =================

>     price                     Float

Reviewer 2: The psolution](https://github.com/odoo/technical-training-solutions/blob/14.0-core/estate/models/estate_property_offer.py#L23) has a `required=True`; who is right?
